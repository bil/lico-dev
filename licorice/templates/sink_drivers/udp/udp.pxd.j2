from runner_utils cimport times_t
from sink_drivers cimport sink_driver

cdef class UdpSinkDriver(sink_driver.SinkDriver):
    cdef int s
    cdef sockaddr_in si_other
    cdef socklen_t slen

    cdef void run(self, times_t *times, void *outBuf, size_t outBufLen, object in_sigs, object in_sig_lens) except *
    cdef void exit_handler(self, int exitStatus) except *
