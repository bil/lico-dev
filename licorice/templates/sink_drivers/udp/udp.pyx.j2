from libc.stdint cimport uint8_t

from sink_drivers cimport sink_driver

cdef extern from "<arpa/inet.h>":
    uint16_t htons(uint16_t hostshort)
    int inet_aton(const char *cp, in_addr *inp)

cdef extern from "<netinet/in.h>":
    enum: IPPROTO_UDP
    ctypedef uint16_t in_port_t
    ctypedef unsigned short sa_family_t
    struct in_addr:
        uint32_t s_addr
    struct sockaddr_in:
        sa_family_t sin_family
        in_port_t sin_port
        in_addr sin_addr
        unsigned char sin_zero[8]

cdef extern from "<sys/socket.h>":
    enum: AF_INET
    enum: SOCK_DGRAM
    ctypedef uint32_t socklen_t
    struct sockaddr:
        unsigned char sa_len
        unsigned char sa_family
        char  sa_data[14]
    ssize_t recvfrom(int, void *restrict, size_t, int, sockaddr *restrict, socklen_t *restrict)
    int bind(int, const sockaddr *, socklen_t)
    int socket(int domain, int type, int protocol)
    ssize_t sendto(int socket, const void *message, size_t length,
             int flags, const sockaddr *dest_addr,
             socklen_t dest_len)


cdef class UdpSinkDriver(sink_driver.SinkDriver):
    def __cinit__(self):
        self.slen = sizeof(self.si_other)

        # create a UDP socket
        self.s = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
        if (self.s == -1):
            die("NETWORK ERROR: socket\n")
        # zero out the structure
        memset(<char *> &self.si_other, 0, sizeof(self.si_other))
        self.si_other.sin_family = AF_INET
        self.si_other.sin_port = htons({{out_signal['args']['port']}})
        if (inet_aton("{{out_signal['args']['addr']}}", &self.si_other.sin_addr)==0):
            die("inet_aton() failed\n")


    cdef void run(self, uint8_t *outBuf, size_t outBufLen, object in_sigs, object in_sig_lens) except *
        # TODO read in MTU or make constant for user. or could use sendto max
        for sentBytes in range(0, outBufLen, 1472):
            if (sendto(self.s, outBuf + sentBytes, 1472, 0, <sockaddr *>&self.si_other, self.slen)==-1):
                die("sendto() failed")


    cdef void exit_handler(self, int exitStatus) except *
        close(self.s)
