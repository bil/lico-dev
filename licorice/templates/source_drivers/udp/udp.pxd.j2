from libc.stdint cimport uint8_t

from source_drivers cimport source_driver

cdef class UdpSourceDriver(source_driver.SourceDriver):
    cdef sockaddr_storage src_addr
    cdef socklen_t src_addr_len

    cdef addrinfo hints
    cdef addrinfo *result, *rp
    cdef int sfd, s
    cdef ssize_t recvLen
    cdef timeval tv
    cdef float sleep_duration

    cdef enum:
        MAX_NUM_PACKETS_PER_TICK = {{in_signal['schema']['max_packets_per_tick']}}


    cdef void run(self, uint8_t **inBuf, size_t *inBufLen, object *out_sigs, object *out_sig_lens) except *
    cdef void exit_handler(self, int exitStatus) except *
