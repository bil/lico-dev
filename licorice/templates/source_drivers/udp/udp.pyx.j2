# __DRIVER_CODE__ imports

from time import sleep
from posix.types cimport suseconds_t, time_t

cdef extern from "<netdb.h>":
    const char *gai_strerror(int ecode);
    struct addrinfo:
        int ai_family
        int ai_socktype
        int ai_flags
        int ai_protocol
        char *ai_canonname
        sockaddr *ai_addr
        addrinfo *ai_next
        int ai_addrlen
    int getaddrinfo (const char *node, const char *service, const addrinfo *hints, addrinfo **res)
    void freeaddrinfo (addrinfo *ai)
    enum: AI_PASSIVE

cdef extern from "<arpa/inet.h>":
    ctypedef uint32_t in_addr_t
    uint16_t htons(uint16_t hostshort)
    in_addr_t inet_addr(const char *cp);

cdef extern from "<netinet/in.h>":
    enum: IPPROTO_UDP
    ctypedef unsigned short sa_family_t

cdef extern from "<sys/socket.h>":
    enum: AF_INET
    enum: SOCK_DGRAM
    enum: MSG_DONTWAIT
    enum: SOL_SOCKET
    enum: SO_RCVTIMEO
    struct sockaddr:
        unsigned char sa_len
        unsigned char sa_family
        char  sa_data[14]
    ssize_t recvfrom(int, void *restrict, size_t, int, sockaddr *restrict, socklen_t *restrict)
    int bind(int, const sockaddr *, socklen_t)
    int socket(int domain, int type, int protocol)
    int setsockopt(int socket, int level, int option_name,
             const void *option_value, socklen_t option_len);

    cdef struct sockaddr_storage:
            unsigned char sa_len
            sa_family_t sa_family

cdef extern from "<sys/time.h>":
    cdef struct timeval:
        time_t tv_sec
        suseconds_t tv_usec


cdef class UdpSourceDriver(source_driver.SourceDriver):
    def __cinit__(self):
        self.src_addr_len = sizeof(sockaddr_storage)

        # create a UDP socket
        memset(&self.hints, 0, sizeof(addrinfo))
        self.hints.ai_family = AF_INET       # IPv4 only
        self.hints.ai_socktype = SOCK_DGRAM  # Datagram socket
        self.hints.ai_flags = AI_PASSIVE     # For use with bind
        self.hints.ai_protocol = IPPROTO_UDP # UDP only
        self.hints.ai_canonname = NULL
        self.hints.ai_addr = NULL
        self.hints.ai_next = NULL

        self.s = getaddrinfo(
            "{{in_signal['args']['interface']}}",
            "{{in_signal['args']['port']}}",
            &self.hints,
            &self.result
        )
        if (self.s != 0):
            fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(s))
            die("getaddrinfo")

        self.rp = result
        while (self.rp != NULL):
            self.sfd = socket(self.rp.ai_family, self.rp.ai_socktype, self.rp.ai_protocol)
            if (self.sfd == -1):
                continue
            if (bind(self.sfd, self.rp.ai_addr, self.rp.ai_addrlen) == 0):
                break # Success
            close(self.sfd)
            self.rp = self.rp.ai_next

        if (self.rp == NULL): # No address succeeded
            fprintf(stderr, "Could not bind\n")
            die("Could not bind")


        self.tv.tv_sec = 60
        self.tv.tv_usec = 0
        if (setsockopt(self.sfd, SOL_SOCKET, SO_RCVTIMEO, &self.tv, sizeof(self.tv)) < 0):
            die("Could not set socket timeout.")

        freeaddrinfo(self.result) # No longer needed

        self.sleep_duration = {{config["config"]["tick_len"]}} / (2. * 1e6)

    cdef void run(self, uint8_t **inBuf, size_t *inBufLen, object *out_sigs, object *out_sig_lens) except *

        if (recvfrom(self.sfd, <void *>inBuf[0], self.packetSize, 0, <sockaddr *> &self.src_addr, &self.src_addr_len) == -1):
            if (errno != EINTR):
                die("NETWORK ERROR: recvfrom()\n")
            else:
                die("Read error.\n")

# {%- if async %}
        #   sleep(self.sleep_duration)
# {%- endif %}


    cdef void exit_handler(self, int exitStatus) except *
        if (self.sfd):
            close(self.sfd)
