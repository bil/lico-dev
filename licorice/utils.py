import os


def __handle_completed_process(result, print_stdout=False):
    if print_stdout:
        print(result.stdout.decode())

    if result.returncode != 0:
        print(f"Completed process output: {result.stdout.decode()}")
        raise RuntimeError(result.stderr.decode())


# return valid filepath if file exists in given path_list
def __find_in_path(path_list, file, raise_error=True):
    for path in path_list:
        filepath = os.path.join(path, file)
        if os.path.exists(filepath):
            return filepath

    if raise_error:
        raise FileNotFoundError(
            f"Could not locate {file} in path: {os.pathsep.join(path_list)}."
        )
    else:
        return None
