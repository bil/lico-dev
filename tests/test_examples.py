import os
import sqlite3
import subprocess
from collections import OrderedDict

import msgpack
import parallel
import pygame
import pytest
from udp import udp_server_recv

from tests.utils import TIME_COLNAMES, validate_model_output


def num_parports():
    num_ports = 0
    for i in range(4):
        try:
            parallel.Parallel(port=i)
            num_ports += 1
        except FileNotFoundError:
            pass

    return num_ports


def has_joystick():
    pygame.joystick.init()
    return pygame.joystick.get_count() >= 1


# TODO create better test that senses if ESP32 is connected
@pytest.mark.skipif(
    num_parports() < 1, reason="No parallel ports connected."
)
def test_jitter(capfd):
    num_ticks = 1000
    subprocess.call(
        f"""{pytest.examples_dir}/jitter/run.sh """
        f"""--config '{{"config":{{"num_ticks":{num_ticks}}}}}'""",
        shell=True,
    )

    validate_model_output(capfd, num_ticks)

    try:
        mean, std_dev, min_val, max_val = udp_server_recv(3333, timeout=1.)
        results_str = f"{mean},{std_dev},{min_val},{max_val}\n"
    except Exception as e:
        print(e)
        results_str = None

    if results_str:
        write_header = False
        if not os.path.exists(f"{pytest.test_dir}/jitter_results.csv"):
            write_header = True

        with open(f"{pytest.test_dir}/jitter_results.csv", "w") as f:
            if write_header:
                f.write("mean,std_dev,min_val,max_val\n")
            f.write(results_str)


@pytest.mark.skipif(
    num_parports() < 2, reason="Insufficient parallel ports connected."
)
def test_parallel_toggle(capfd):
    num_ticks = 1000
    subprocess.call(
        f"""{pytest.examples_dir}/parallel_toggle/run.sh """
        f"""--config '{{"config":{{"num_ticks":{num_ticks}}}}}'""",
        shell=True,
    )

    validate_model_output(capfd, num_ticks)


def test_matrix_multiply(capfd):
    num_ticks = 30
    subprocess.call(
        f"{pytest.examples_dir}/matrix_multiply/run.sh "
        f"""--config '{{"config":{{"num_ticks":{num_ticks}}}}}'""",
        shell=True
    )

    validate_model_output(capfd, num_ticks)


def test_logger(capfd, snapshot):
    num_ticks = 32
    os.environ["SDL_VIDEODRIVER"] = "dummy"
    subprocess.call(
        f"{pytest.examples_dir}/logger/run.sh "
        f"""--config '{{"config":{{"num_ticks":{num_ticks}}}}}'""",
        shell=True
    )

    # # TODO address UserWarning for no parser. Move validation into drivers
    validate_model_output(capfd, num_ticks, validate_stderr=False)

    conn = sqlite3.connect(
        f"{pytest.examples_dir}/logger/logger_demo.lico/out/data_0000.db"
    )
    cur = conn.cursor()
    cur.execute("PRAGMA table_info(tick);")
    col_names = [val[1] for val in cur.fetchall()]
    # don't snapshot nanosecond timers
    for n in TIME_COLNAMES[1:]:
        col_names.remove(n)

    cur.execute(f"SELECT {' ,'.join(col_names)} FROM tick;")

    col_names = [col[0] for col in cur.description]
    col_vals = [[] for _ in col_names]
    for row in cur.fetchall():
        for i in range(len(col_names)):
            val = row[i]
            if col_names[i][0] == "m":
                val = msgpack.unpackb(val)
            col_vals[i].append(val)
    table_dict = OrderedDict(sorted(dict(zip(col_names, col_vals)).items()))
    assert table_dict == snapshot


@pytest.mark.skipif(
    not has_joystick(), reason="No joystick connected."
)
def test_joystick(capfd):
    num_ticks = 21
    os.environ["SDL_VIDEODRIVER"] = "dummy"
    subprocess.call(f"{pytest.examples_dir}/joystick/run.sh", shell=True)

    validate_model_output(capfd, num_ticks)


def test_pygame(capfd):
    num_ticks = 100
    os.environ["SDL_VIDEODRIVER"] = "dummy"
    subprocess.call(
        f"{pytest.examples_dir}/pygame/run.sh "
        f"""--config '{{"config":{{"num_ticks":{num_ticks}}}}}'""",
        shell=True
    )

    validate_model_output(capfd, num_ticks)


@pytest.mark.skipif(
    not has_joystick(), reason="No joystick connected."
)
def test_cursor_track(capfd):
    num_ticks = 100
    os.environ["SDL_VIDEODRIVER"] = "dummy"
    subprocess.call(
        f"{pytest.examples_dir}/cursor_track/run.sh "
        f"""--config '{{"config":{{"num_ticks":{num_ticks}}}}}'""",
        shell=True
    )

    # TODO address UserWarning for no parser. Move validation into drivers
    validate_model_output(capfd, num_ticks, validate_stderr=False)


@pytest.mark.skipif(
    not has_joystick(), reason="No joystick connected."
)
def test_pinball(capfd):
    num_ticks = 100
    os.environ["SDL_VIDEODRIVER"] = "dummy"
    subprocess.call(
        f"{pytest.examples_dir}/pinball/run.sh "
        f"""--config '{{"config":{{"num_ticks":{num_ticks}}}}}'""",
        shell=True
    )

    # TODO address UserWarning for no parser. Move validation into drivers
    validate_model_output(capfd, num_ticks, validate_stderr=False)


def test_udp(capfd):
    proc = subprocess.Popen([f"{pytest.examples_dir}/udp/send_udp.sh"])

    subprocess.call(
        f"""{pytest.examples_dir}/udp/run.sh """
        f"""--config '{{"config":{{"num_ticks":{600}}}}}'""",
        shell=True,
    )

    proc.terminate()

    captured = capfd.readouterr()
    print(captured.out)
    print(captured.err)
    msgs = ["STAN", "FORD", "TREE", "MESS", "AGES"]
    for msg in msgs:
        assert msg in captured.out


@pytest.mark.xfail(raises=NotImplementedError)
def test_sine(capfd):
    raise NotImplementedError


@pytest.mark.xfail(raises=NotImplementedError)
def test_audio_capture(capfd):
    raise NotImplementedError


@pytest.mark.xfail(raises=NotImplementedError)
def test_polaris(capfd):
    raise NotImplementedError
