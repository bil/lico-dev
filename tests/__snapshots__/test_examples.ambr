# name: test_logger
  OrderedDict({
    'm_f8_matrix_out': list([
      list([
        3.0,
        3.0,
        3.0,
        3.0,
      ]),
      list([
        9.0,
        9.0,
        9.0,
        9.0,
      ]),
      list([
        27.0,
        27.0,
        27.0,
        27.0,
      ]),
      list([
        81.0,
        81.0,
        81.0,
        81.0,
      ]),
      list([
        243.0,
        243.0,
        243.0,
        243.0,
      ]),
      list([
        729.0,
        729.0,
        729.0,
        729.0,
      ]),
      list([
        2187.0,
        2187.0,
        2187.0,
        2187.0,
      ]),
      list([
        6561.0,
        6561.0,
        6561.0,
        6561.0,
      ]),
      list([
        19683.0,
        19683.0,
        19683.0,
        19683.0,
      ]),
      list([
        59049.0,
        59049.0,
        59049.0,
        59049.0,
      ]),
      list([
        177147.0,
        177147.0,
        177147.0,
        177147.0,
      ]),
      list([
        531441.0,
        531441.0,
        531441.0,
        531441.0,
      ]),
      list([
        1594323.0,
        1594323.0,
        1594323.0,
        1594323.0,
      ]),
      list([
        4782969.0,
        4782969.0,
        4782969.0,
        4782969.0,
      ]),
      list([
        14348907.0,
        14348907.0,
        14348907.0,
        14348907.0,
      ]),
      list([
        43046721.0,
        43046721.0,
        43046721.0,
        43046721.0,
      ]),
      list([
        129140163.0,
        129140163.0,
        129140163.0,
        129140163.0,
      ]),
      list([
        387420489.0,
        387420489.0,
        387420489.0,
        387420489.0,
      ]),
      list([
        1162261467.0,
        1162261467.0,
        1162261467.0,
        1162261467.0,
      ]),
      list([
        3486784401.0,
        3486784401.0,
        3486784401.0,
        3486784401.0,
      ]),
      list([
        10460353203.0,
        10460353203.0,
        10460353203.0,
        10460353203.0,
      ]),
      list([
        1.0,
        2.0,
        3.0,
        4.0,
      ]),
      list([
        3.0,
        6.0,
        9.0,
        12.0,
      ]),
      list([
        9.0,
        18.0,
        27.0,
        36.0,
      ]),
      list([
        27.0,
        54.0,
        81.0,
        108.0,
      ]),
      list([
        81.0,
        162.0,
        243.0,
        324.0,
      ]),
      list([
        243.0,
        486.0,
        729.0,
        972.0,
      ]),
      list([
        729.0,
        1458.0,
        2187.0,
        2916.0,
      ]),
      list([
        2187.0,
        4374.0,
        6561.0,
        8748.0,
      ]),
      list([
        6561.0,
        13122.0,
        19683.0,
        26244.0,
      ]),
      list([
        19683.0,
        39366.0,
        59049.0,
        78732.0,
      ]),
      list([
        59049.0,
        118098.0,
        177147.0,
        236196.0,
      ]),
    ]),
    'r_i1_scalar_out': list([
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
      32,
      33,
    ]),
    'time_tick': list([
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
    ]),
    'v_f8_vector_floats_x': list([
      2.0,
      4.0,
      8.0,
      16.0,
      32.0,
      64.0,
      128.0,
      256.0,
      512.0,
      1024.0,
      2048.0,
      4096.0,
      8192.0,
      16384.0,
      32768.0,
      65536.0,
      131072.0,
      262144.0,
      524288.0,
      1048576.0,
      2097152.0,
      4194304.0,
      8388608.0,
      16777216.0,
      33554432.0,
      67108864.0,
      134217728.0,
      268435456.0,
      536870912.0,
      1073741824.0,
      1.0,
      2.0,
    ]),
    'v_f8_vector_floats_y': list([
      2.0,
      4.0,
      8.0,
      16.0,
      32.0,
      64.0,
      128.0,
      256.0,
      512.0,
      1024.0,
      2048.0,
      4096.0,
      8192.0,
      16384.0,
      32768.0,
      65536.0,
      131072.0,
      262144.0,
      524288.0,
      1048576.0,
      2097152.0,
      4194304.0,
      8388608.0,
      16777216.0,
      33554432.0,
      67108864.0,
      134217728.0,
      268435456.0,
      536870912.0,
      1073741824.0,
      2.0,
      4.0,
    ]),
    'v_f8_vector_floats_z': list([
      2.0,
      4.0,
      8.0,
      16.0,
      32.0,
      64.0,
      128.0,
      256.0,
      512.0,
      1024.0,
      2048.0,
      4096.0,
      8192.0,
      16384.0,
      32768.0,
      65536.0,
      131072.0,
      262144.0,
      524288.0,
      1048576.0,
      2097152.0,
      4194304.0,
      8388608.0,
      16777216.0,
      33554432.0,
      67108864.0,
      134217728.0,
      268435456.0,
      536870912.0,
      1073741824.0,
      3.0,
      6.0,
    ]),
    'v_f8_vector_floats_z2': list([
      2.0,
      4.0,
      8.0,
      16.0,
      32.0,
      64.0,
      128.0,
      256.0,
      512.0,
      1024.0,
      2048.0,
      4096.0,
      8192.0,
      16384.0,
      32768.0,
      65536.0,
      131072.0,
      262144.0,
      524288.0,
      1048576.0,
      2097152.0,
      4194304.0,
      8388608.0,
      16777216.0,
      33554432.0,
      67108864.0,
      134217728.0,
      268435456.0,
      536870912.0,
      1073741824.0,
      4.0,
      8.0,
    ]),
    'v_i2_vector_ints_0': list([
      -32767,
      -32762,
      -32757,
      -32752,
      -32747,
      -32742,
      -32737,
      -32732,
      -32727,
      -32722,
      -32717,
      -32712,
      -32707,
      -32702,
      -32697,
      -32692,
      -32687,
      -32682,
      -32677,
      -32672,
      -32667,
      -32662,
      -32657,
      -32652,
      -32647,
      -32642,
      -32637,
      -32632,
      -32627,
      -32622,
      -32617,
      -32612,
    ]),
    'v_i2_vector_ints_1': list([
      -20000,
      -19995,
      -19990,
      -19985,
      -19980,
      -19975,
      -19970,
      -19965,
      -19960,
      -19955,
      -19950,
      -19945,
      -19940,
      -19935,
      -19930,
      -19925,
      -19920,
      -19915,
      -19910,
      -19905,
      -19900,
      -19895,
      -19890,
      -19885,
      -19880,
      -19875,
      -19870,
      -19865,
      -19860,
      -19855,
      -19850,
      -19845,
    ]),
    'v_i2_vector_ints_2': list([
      -10000,
      -9995,
      -9990,
      -9985,
      -9980,
      -9975,
      -9970,
      -9965,
      -9960,
      -9955,
      -9950,
      -9945,
      -9940,
      -9935,
      -9930,
      -9925,
      -9920,
      -9915,
      -9910,
      -9905,
      -9900,
      -9895,
      -9890,
      -9885,
      -9880,
      -9875,
      -9870,
      -9865,
      -9860,
      -9855,
      -9850,
      -9845,
    ]),
  })
# ---
